Mathlib/Topology/CWComplex.lean

```diff
@@ -0,0 +1,140 @@
+indexed by `cells` -/
+def generalizedSigmaSphereInclusion (n : ℤ) (cells : Type) :
+    TopCat.of (Σ (_ : cells), S n) ⟶ TopCat.of (Σ (_ : cells), D (n + 1)) where
+  toFun := Sigma.map id fun _ x ↦ (f n).toFun x
+  continuous_toFun := Continuous.sigma_map fun _ ↦ (f n).continuous_toFun
+
+/-- Given an attaching map for each `S n` (boundary of a generalized `(n+1)`-cell), we construct
+the attaching map for the disjoint union of all the `S n`. -/
+def generalizedSigmaAttachMap (X : TopCat.{u}) (n : ℤ) (cells : Type)
+    (attach_maps : cells → C(S n, X)) : TopCat.of (Σ (_ : cells), S n) ⟶ X where
+  toFun := fun ⟨i, x⟩ ↦ attach_maps i x
+  continuous_toFun := continuous_sigma fun i ↦ (attach_maps i).continuous_toFun
+
+/-- A type witnessing that `X'` is obtained from `X` by attaching generalized `(n+1)`-cells, where
+a generalized `(n+1)`-cell is given by `f n : S n ⟶ D (n + 1)`. -/
+structure AttachGeneralizedCells (X X' : TopCat.{u}) (n : ℤ) where
+  /-- The index type over the generalized `(n+1)`-cells -/
+  cells : Type
+  /-- For each generalized `(n+1)`-cell, we have an attaching map from its boundary to `X`. -/
+  attach_maps : cells → C(S n, X)
+  /-- `X'` is the pushout obtained from `X` along `sigmaAttachMap`. -/
+  iso_pushout : X' ≅ Limits.pushout (generalizedSigmaSphereInclusion f n cells)
+    (generalizedSigmaAttachMap X n cells attach_maps)
+
+/-- The inclusion map from the disjoint union of `n`-spheres to the disjoint union of `(n+1)`-disks,
+where both of the disjoint unions are indexed by `cells` -/
+noncomputable abbrev sigmaSphereInclusion := generalizedSigmaSphereInclusion sphereInclusion
+
+/-- Given an attaching map for each `n`-sphere, we construct the attaching map for the disjoint
+union of the `n`-spheres. -/
+abbrev sigmaAttachMap := @generalizedSigmaAttachMap sphere
+
+/-- A type witnessing that `X'` is obtained from `X` by attaching `(n+1)`-disks -/
+abbrev AttachCells := AttachGeneralizedCells sphereInclusion
+
+end RelativeCWComplex
+
+/-- A relative CW-complex contains an expanding sequence of subspaces `sk i` (called the
+`(i-1)`-skeleton) for `i ≥ 0`, where `sk 0` (i.e., the `(-1)`-skeleton) is an arbitrary topological
+space, and each `sk (n+1)` (i.e., the `n`-skeleton) is obtained from `sk n` (i.e., the
+`(n-1)`-skeleton) by attaching `n`-disks. -/
+structure RelativeCWComplex where
+  /-- The skeletons. Note: `sk i` is usually called the `(i-1)`-skeleton in the math literature. -/
+  sk : ℕ → TopCat.{u}
+  /-- Each `sk (n+1)` (i.e., the `n`-skeleton) is obtained from `sk n` (i.e., the
+  `(n-1)`-skeleton) by attaching `n`-disks. -/
+  attach_cells : (n : ℕ) → RelativeCWComplex.AttachCells (sk n) (sk (n + 1)) (n - 1)
+
+/-- A CW-complex is a relative CW-complex whose `sk 0` (i.e., `(-1)`-skeleton) is empty. -/
+structure CWComplex extends RelativeCWComplex.{u} where
+  /-- `sk 0` (i.e., the `(-1)`-skeleton) is empty. -/
+  sk_zero_empty : sk 0 = TopCat.of (ULift Empty)
```

Comment:
This looks really wrong in that you are stating equality of types. Is there any reason you are not doing
```suggestion
  sk_zero_empty : IsEmpty (sk 0)
```

Reply:
`IsEmpty` indeed makes proofs simpler. Thanks!